app/com/linkedin/drelephant/spark/aggregator/YarnSparkMetricsAggregator.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2016 LinkedIn Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.linkedin.drelephant.spark.aggregator;
+
+import com.linkedin.drelephant.analysis.HadoopAggregatedData;
+import com.linkedin.drelephant.analysis.HadoopApplicationData;
+import com.linkedin.drelephant.analysis.HadoopMetricsAggregator;
+import com.linkedin.drelephant.configurations.aggregator.AggregatorConfigurationData;
+import com.linkedin.drelephant.math.Statistics;
+import com.linkedin.drelephant.spark.data.SparkApplicationData;
+import com.linkedin.drelephant.spark.data.SparkExecutorData;
+import com.linkedin.drelephant.util.SparkUtils;
+import org.apache.commons.io.FileUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Iterator;
+
+/**
+ * A more general and accurate implementation of SparkMetricsAggregator which is compatible with Spark on YARN
+ * as well as executor dynamic allocation mode.
+ *
+ * What's new:
+ * 1) Memory: Use actual executor memory request considering `overhead` instead of only `spark.executor.memory`.
+ * 2) Time: Use running time of all existed executor, not duration(sum of task executing time) of
+ *      executors only including of those which are alive while ending of application.
+ *
+ */
+public class YarnSparkMetricsAggregator implements HadoopMetricsAggregator {
+
+  private static final Logger logger = LoggerFactory.getLogger(YarnSparkMetricsAggregator.class);
+
+  private AggregatorConfigurationData _aggregatorConfigurationData;
+  private double _storageMemWastageBuffer = 0.5;
```

Comment:
I thought this variable will be useless in Resource aggregation of Spark, because like MapReduce, we should figure out way of identifying the `actual physical memory usage` of each Executor, then use `(1-physicalMemoryUsage/ExecutorMemory)` as resource wasted. But in our company we add feature of checking the `actual physical memory usage` embedded in our Spark branch(different from official spark version).
That's some suggestion about `resource wasted` metric calculation, and i will change `_storageMemwastageBuffer` for temporary.